astropy/io/misc/hdf5.py

```diff
@@ -282,10 +296,9 @@ def write_table_hdf5(table, output, path=None, compression=False,
 
         # Open the file for appending or writing
         f = h5py.File(output, 'a' if append else 'w')
-
         # Recursively call the write function
         try:
-            return write_table_hdf5(table, f, path=path,
+            return write_table_hdf5(check_column_encoding(table), f, path=path,
```

Comment:
I'm not familiar with the hdf5 format not with this function, but given the comment above ("Recursively call the write...") and the possibility to write to a h5py object, I think the encoding should not be done here, but probably below, near the `_encode_mixins` line. Also it may be useful to add a test with a file object (and/or h5py file object ?).